File: arrays/problems/java/ReverseString/src/test/java/org/data/problem/ReverseStringTest.java
package org.data.problem;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

public class ReverseStringTest {

    @Test
    public void testEmptyString() {
        String cad = "";

        String actual = ReverseString.calculate(cad);
        assertEquals(actual, cad, "We expect an empty string");
    }

    @Test
    public void testNullString() {
        String actual = ReverseString.calculate(null);
        assertNull(actual, "We expect a null string");
    }

    @Test
    public void testStringOneChar() {
        String cad = "a";

        String actual = ReverseString.calculate(cad);
        assertEquals(actual, cad, "We expect a string equal to a");
    }

    @Test
    public void testStringOneWord() {
        String cad = "skate";

        String actual = ReverseString.calculate(cad);
        assertEquals(actual, "etaks", "We expect a string equal to etaks");
    }

    @Test
    public void testStringOneWordWithSpaces() {
        String cad = " skate ";

        String actual = ReverseString.calculate(cad);
        assertEquals(actual, " etaks ", "We expect a string equal to etaks with spaces");
    }

    @Test
    public void testStringMultipleWords() {
        String cad = "Inline skates is cool";

        String actual = ReverseString.calculate(cad);
        assertEquals(actual, "looc si setaks enilnI", "We expect a string equal to etaks with spaces");
    }
}

File: arrays/problems/go/ReverseString/ReverseString_test.go
package main

import "testing"

func TestEmptyString(t *testing.T) {
	actual := ReverseString("")
	if actual != "" {
		t.Errorf("Actual %s is different than expected %s", actual, "")
	}

	if len(actual) != 0 {
		t.Errorf("Length %d is different to %d", len(actual), 0)
	}
}

func TestStringOneChar(t *testing.T) {
	expected := "a"
	actual := ReverseString(expected)

	if actual != expected {
		t.Errorf("Actual %s is different than expected %s", actual, expected)
	}

	if len(actual) != len(expected) {
		t.Errorf("Length %d is different to %d", len(actual), len(expected))
	}
}

func TestStringOneWord(t *testing.T) {
	word := "skate"
	expected := "etaks"
	actual := ReverseString(word)

	if actual != expected {
		t.Errorf("Actual %s is different than expected %s", actual, expected)
	}

	if len(actual) != len(expected) {
		t.Errorf("Length %d is different to %d", len(actual), len(expected))
	}
}

func TestStringOneWordWithSpaces(t *testing.T) {
	word := " skate "
	expected := " etaks "
	actual := ReverseString(word)

	if actual != expected {
		t.Errorf("Actual %s is different than expected %s", actual, expected)
	}

	if len(actual) != len(expected) {
		t.Errorf("Length %d is different to %d", len(actual), len(expected))
	}
}

func TestStringMultipleWords(t *testing.T) {
	word := "Inline skates is cool"
	expected := "looc si setaks enilnI"
	actual := ReverseString(word)

	if actual != expected {
		t.Errorf("Actual %s is different than expected %s", actual, expected)
	}

	if len(actual) != len(expected) {
		t.Errorf("Length %d is different to %d", len(actual), len(expected))
	}
}
